memguardian/memguardian.go

```diff
@@ -0,0 +1,106 @@
+package memguardian
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+	"time"
+
+	"github.com/shirou/gopsutil/mem"
+)
+
+type MemGuardianOption func(*MemGuardian) error
+
+// WithInterval defines the ticker interval of the memory monitor
+func WitInterval(d time.Duration) MemGuardianOption {
+	return func(mg *MemGuardian) error {
+		mg.t = time.NewTicker(d)
+		return nil
+	}
+}
+
+// WithCallback defines an optional callback if the warning ration is exceeded
+func WithCallback(f func()) MemGuardianOption {
+	return func(mg *MemGuardian) error {
+		mg.f = f
+		return nil
+	}
+}
+
+// WithRatioWarning defines the threshold of the warning state (and optional callback invocation)
+func WithRatioWarning(ratio float64) MemGuardianOption {
+	return func(mg *MemGuardian) error {
+		if ratio == 0 || ratio > 100 {
+			return errors.New("ratio must be between 1 and 100")
+		}
+		mg.ratio = ratio
+		return nil
+	}
+}
+
+type MemGuardian struct {
+	t       *time.Ticker
+	f       func()
+	ctx     context.Context
+	cancel  context.CancelFunc
+	Warning atomic.Bool
+	ratio   float64
+}
+
+// New mem guadian instance with user defined options
+func New(options ...MemGuardianOption) (*MemGuardian, error) {
+	mg := &MemGuardian{}
+	for _, option := range options {
+		if err := option(mg); err != nil {
+			return nil, err
+		}
+	}
+
+	mg.ctx, mg.cancel = context.WithCancel(context.TODO())
+
+	return mg, nil
+}
+
+// Run the instance monitor (cancel using the Stop method or context parameter)
+func (mg *MemGuardian) Run(ctx context.Context) error {
+	for {
+		select {
+		case <-mg.ctx.Done():
+			mg.Close()
+			return nil
+		case <-ctx.Done():
+			mg.Close()
+			return nil
+		case <-mg.t.C:
+			usedRatio, err := UsedRamRatio()
+			if err != nil {
+				return err
+			}
+
+			if usedRatio >= mg.ratio {
+				mg.Warning.Store(true)
+				if mg.f != nil {
+					mg.f()
+				}
+			} else {
+				mg.Warning.Store(false)
+			}
+		}
+	}
+}
+
+// Close and stops the instance
+func (mg *MemGuardian) Close() {
+	mg.cancel()
+	mg.t.Stop()
+}
+
+// Calculate the system absolute ratio of used RAM vs total available (as of now doesn't consider swap)
+func UsedRamRatio() (float64, error) {
+	vms, err := mem.VirtualMemory()
```

Comment:
Are you sure that the virtual memory is the right metric? Virtual memory can be larger than the physical available memory. How about considering the real memory or physical memory?

Reply:
I thought about retrieving directly via OS system call (eg. [sysinfo](https://man7.org/linux/man-pages/man2/sysinfo.2.html) but the [gopsutil](https://github.com/shirou/gopsutil) seems to do already a decent job about obtaining it with standard system tools/files (for example on linux by parsing `/proc/meminfo` which according to linux [kernel docs](https://www.kernel.org/doc/Documentation/filesystems/proc.txt) represent the avilable effective RAM). Do you think this is correct, or there is any better way on the metrics to monitor (for example swap is actually not considered)?

Reply:
This will probably differ per OS:
On MacOS Virtual means something else:
https://apple.stackexchange.com/a/107

Additionally misleading info from gopls:
https://github.com/shirou/gopsutil/issues/119

Windows is probably the same

Reply:
Thanks for the links, I'll have a closer look and see if it makes sense to implement more specific os calls